dohproxy/server_protocol.py

```diff
@@ -40,6 +40,7 @@ def __init__(self, upstream_resolver, upstream_port, logger=None):
         if logger is None:
             logger = utils.configure_logger('DNSClient', 'DEBUG')
         self.logger = logger
+        self.transport = None
```

Comment:
Hmm, storing `transport` on the `DNSClient` object only works here because you create a new instance of `DNSClient` before every call to `query`. This wouldn't be safe if multiple callers ran `await instance.query(` on the same `DNSClient` object. I think this won't break this repository, but won't play nice with anyone assuming `query` can be called asynchronously multiple times in "parallel".

Reply:
that's right. Indeed, this is an assumption made in this repo: https://github.com/facebookexperimental/doh-proxy/blob/master/dohproxy/proxy.py#L196 but I definitely see your point and it is totally fair.

Do you use it as a library? or do you use the tools?

Reply:
I came across your library from https://github.com/rthalley/dnspython/issues/370 when I had been learning the steps to implement an asyncio.Protocol myself. I learned a lot of the edge cases from reading this implementation, but then in extensive testing of my own code I ran across the missing `transport.close()` case when timing out queries. I figured I'd share my learnings back here :smiley:

Reply:
My actual implementation is a blend of the asyncio.Protocol from here, the query logic from https://github.com/rthalley/dnspython/blob/master/dns/resolver.py#L804 and the event buffering from https://github.com/aio-libs/aiohttp/blob/48bf8c356c7c3bac5d520484ede5c35bd1795afb/aiohttp/connector.py#L788-L798

Reply:
> I came across your library from rthalley/dnspython#370 when I had been learning the steps to implement an asyncio.Protocol myself. 

Yeah, I wished there were a solution worked on in upstream so we would not need to have different half-baked implementations.

> I learned a lot of the edge cases from reading this implementation, but then in extensive testing of my own code I ran across the missing transport.close() case when timing out queries. I figured I'd share my learnings back here 😃

Thanks @Peter200lx, really appreciated! Hopefully those learnings can be re-used the day dnspython start having an implementation.